xdsl/dialects/dlti.py

```diff
@@ -47,12 +71,39 @@ class DLTIEntryMap(ParametrizedAttribute, ABC):
     # In MLIR, this is a DataLayoutEntryInterface.
     entries: ArrayAttr[DataLayoutEntryAttr]
 
+    def __init__(
+        self,
+        contents: ArrayAttr
+        | list[DataLayoutEntryAttr]
+        | dict[StringAttr | TypeAttribute | str, Attribute | str | int | float],
+    ):
+        if isinstance(contents, dict):
+            contents = ArrayAttr(
+                [DataLayoutEntryAttr(k, v) for k, v in contents.items()]
+            )
+        elif isinstance(contents, list):
+            contents = ArrayAttr(contents)
+
+        assert isinstance(contents, ArrayAttr)
+
+        super().__init__(contents)
+
     @classmethod
     def parse_parameters(
         cls, parser: AttrParser
     ) -> tuple[ArrayAttr[DataLayoutEntryAttr]]:
         def parse_entry() -> DataLayoutEntryAttr:
+            pos = parser.pos
             entry = parser.parse_attribute()
+            end_pos = parser.pos
+            if not (
+                isinstance(entry, str)
+                or isinstance(entry, StringAttr)
+                or isinstance(entry, TypeAttribute)
+            ):
+                parser.raise_error(
+                    "key must be a string or a type attribute", pos, end_pos
+                )
```

Comment:
If it can only be a StringAttr or TypeAttribute then we should constrain it at the point of declaration instead of in the parser

Reply:
if you added this check for type reasons then I would actually recommend using `DataLayoutEntryAttr.new(entry, value)` below

Reply:
I've removed the check here and use a cast to cast the Attribute returned from parser.parse_attribute into StringAttr | TypeAttribute to keep pyright happy